### PR TITLE
[#15] Bindings for Hex and Base64 encoding/decoding.

### PR DESCRIPTION
**Pull request by sethlivy**
This PR adds bindings to libsodium's serialization utilities along with the constant-time comparison function.

I've built this on my own machine, but I've never written C bindings before, so I don't have a clue if I'm doing it right.

The associated GHC issue that had been blocking this has been closed and backported to recent GHC versions. (at least from what I read)

### CODE REVIEW

File: src/LibSodium/Bindings/Utils.hs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+-- | Constant-time comparison function.
+--
+-- This function is not a lexicographic comparator and should be never
+-- used for this purpose. It should only be used when comparing two pieces
+-- of secret data, such as keys or authentication tags.
+foreign import capi "sodium.h sodium_memcmp"
+  sodiumMemcmp
+    :: Ptr ()
+    -- ^ First pointer to some secret data.
+    -> Ptr ()
```
Comment (kozross) [inline src/LibSodium/Bindings/Utils.hs:43]: `()` is _not_ the same as C's `void`. If anything, this should probably be `Ptr CUChar`.

**Review by kozross — CHANGES_REQUESTED** [review]

File: src/LibSodium/Bindings/Utils.hs (hunk 1 above)
Comment (sethlivy) [reply src/LibSodium/Bindings/Utils.hs]: I've changed both of these arguments to `Ptr CUChar`.

**Review by kozross — APPROVED** [review]

**Review by Kleidukos — COMMENTED** [review]
Congratulations, you can now fix the C type warnings! :D https://github.com/haskell-cryptography/libsodium-bindings/actions/runs/3453287531/jobs/5763786083#step:8:1

**Comment by sethlivy** [comment]
Taken a longer look at this. With the C FFI that Haskell provides, we can't represent nested const pointers without dropping the const qualifier and creating errors. 

There is a fix coming for this in GHC 9.6: [https://gitlab.haskell.org/ghc/ghc/-/commit/4f70a8a0b5db49ff249271faefec14bf1421f365](Here.)

File: src/LibSodium/Bindings/Utils.hs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  sodiumBase64VariantOriginal :: CInt
+
+foreign import capi "sodium.h value sodium_base64_VARIANT_ORIGINAL_NO_PADDING"
+  sodiumBase64VariantOriginalNoPadding :: CInt
+
+foreign import capi "sodium.h value sodium_base64_VARIANT_URLSAFE"
+  sodiumBase64VariantURLSafe :: CInt
+
+foreign import capi "sodium.h value sodium_base64_VARIANT_URLSAFE_NO_PADDING"
+  sodiumBase64VariantURLSafeNoPadding :: CInt
```
Comment (kozross) [inline src/LibSodium/Bindings/Utils.hs:172]: These could do with docs I think.

**Review by kozross — CHANGES_REQUESTED** [review]

**Review by kozross — APPROVED** [review]

**Comment by Kleidukos** [comment]
Fantastic, thank you @sethlivy!

**Pull request merged**